src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -0,0 +1,574 @@
+
+  /** Simplify space using the laws, there's no nested union after simplify */
+  def simplify(space: Space): Space = space match {
+    case Kon(tp, spaces) =>
+      val sp = Kon(tp, spaces.map(simplify _))
+      if (sp.params.exists(_ == Empty)) Empty
+      else sp
+    case Or(spaces) =>
+      val set = spaces.map(simplify _).flatMap {
+        case Or(ss) => ss
+        case s => Seq(s)
+      } filter (_ != Empty)
+
+      if (set.isEmpty) Empty
+      else if (set.size == 1) set.toList(0)
+      else Or(set)
+    case Typ(tp, _) =>
+      if (canDecompose(tp) && partitions(tp).isEmpty) Empty
+      else space
+    case _ => space
+  }
+
+  /** Flatten space to get rid of `Or` for pretty print */
+  def flatten(space: Space): List[Space] = space match {
+    case Kon(tp, spaces) =>
+      val flats = spaces.map(flatten _)
+
+      flats.foldLeft(List[Kon]()) { (acc, flat) =>
+        if (acc.isEmpty) flat.map(s => Kon(tp, Nil :+ s))
+        else for (Kon(tp, ss) <- acc; s <- flat) yield Kon(tp, ss :+ s)
+      }
+    case Or(spaces) =>
+      spaces.flatMap(flatten _)
+    case _ => List(space)
+  }
+
+  /** Is `a` a subspace of `b`? Equivalent to `a - b == Empty`, but faster */
+  def subspace(a: Space, b: Space): Boolean = (a, b) match {
+    case (Empty, _) => true
+    case (_, Empty) => false
+    case (Or(ss), _) => ss.forall(subspace(_, b))
+    case (Typ(tp1, _), Typ(tp2, _)) =>
+      isSubType(tp1, tp2)
+    case (Typ(tp1, _), Or(ss)) =>
+      ss.exists(subspace(a, _)) ||
+      (canDecompose(tp1) && subspace(Or(partitions(tp1)), b))
+    case (Typ(tp1, _), Kon(tp2, ss)) =>
+      isSubType(tp1, tp2) && subspace(Kon(tp2, signature(tp2).map(Typ(_, false))), b)
+    case (Kon(tp1, ss), Typ(tp2, _)) =>
+      isSubType(tp1, tp2) ||
+      simplify(a) == Empty ||
+      (isSubType(tp2, tp1) &&
+        canDecompose(tp1) &&
+        subspace(Or(partitions(tp1)), b))
+    case (Kon(_, _), Or(_)) =>
+      simplify(minus(a, b)) == Empty
+    case (Kon(tp1, ss1), Kon(tp2, ss2)) =>
+      isEqualType(tp1, tp2) && ss1.zip(ss2).forall((subspace _).tupled)
+    case (Const(v1, _), Const(v2, _)) => v1 == v2
+    case (Const(_, tp1), Typ(tp2, _)) => isSubType(tp1, tp2)
+    case (Const(_, _), Or(ss)) => ss.exists(subspace(a, _))
+    case (Const(_, _), _) => false
+    case (_, Const(_, _)) => false
+    case (Var(x, _), Var(y, _)) => x == y
+    case (Var(_, tp1), Typ(tp2, _)) => isSubType(tp1, tp2)
+    case (Var(_, _), Or(ss)) => ss.exists(subspace(a, _))
+    case (Var(_, _), _) => false
+    case (_, Var(_, _)) => false
+  }
+
+  /** Intersection of two spaces  */
+  def intersect(a: Space, b: Space): Space = (a, b) match {
+    case (Empty, _) | (_, Empty) => Empty
+    case (_, Or(ss)) => Or(ss.map(intersect(a, _)))
+    case (Or(ss), _) => Or(ss.map(intersect(_, b)))
+    case (Typ(tp1, _), Typ(tp2, _)) =>
+      if (isSubType(tp1, tp2)) a
+      else if (isSubType(tp2, tp1)) b
+      else Empty
```

Comment:
Why is `A intersect B` empty?
If two traits are unrelated they still can have common inhabitants. See below:
```scala
trait A
trait B

sealed class Parent
class Foo extends Parent with A
class Bar extends Parent with B

(null: parent) match {
  case (_: A) | (_: B) =>
}
```